org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java

```diff
@@ -210,7 +213,35 @@ public ReplaceRegion serializeReplacement(EObject obj, SaveOptions options) {
 			throw new IllegalStateException("Cannot replace an obj that has no associated node");
 		}
 		String text = serialize(obj, options);
-		return new ReplaceRegion(node.getTotalOffset(), node.getTotalLength(), text);
+		int replaceRegionLength = calculateReplaceRegionLength(node, text);
+		return new ReplaceRegion(node.getTotalOffset(), replaceRegionLength, text);
 	}
 
+	protected int calculateReplaceRegionLength(ICompositeNode node, String text) {
+		int oldTextLength = node.getTotalLength();
+		int newTextLength = text.length();
+
+		if (newTextLength > oldTextLength) {
+			String remainingText = text.substring(oldTextLength);
+			if (isWhitespace(remainingText) && hiddenNodeFollows(node)) {
+				return newTextLength;
+			}
+		}
+		return oldTextLength;
+	}
+
+	protected boolean hiddenNodeFollows(ICompositeNode node) {
+		if (node.hasNextSibling()) {
+			INode nextSibling = node.getNextSibling();
+			if (nextSibling instanceof ILeafNode) {
```

Comment:
If the next sibling is a composite node, it's first child may still be a hidden node.
If you want to catch those, too, I recommend to use something like nextSibling.getLeafNodes() and check the first node in that list.